internal/handlers/pg/cmd_query.go

```diff
@@ -16,47 +16,71 @@ package pg
 
 import (
 	"context"
+	"errors"
 	"fmt"
-	"time"
+	"strings"
+
+	"github.com/jackc/pgerrcode"
+	"github.com/jackc/pgx/v5/pgconn"
 
 	"github.com/FerretDB/FerretDB/internal/handlers/common"
 	"github.com/FerretDB/FerretDB/internal/handlers/commonerrors"
 	"github.com/FerretDB/FerretDB/internal/types"
+	"github.com/FerretDB/FerretDB/internal/util/lazyerrors"
 	"github.com/FerretDB/FerretDB/internal/util/must"
 	"github.com/FerretDB/FerretDB/internal/wire"
 )
 
 // CmdQuery implements HandlerInterface.
 func (h *Handler) CmdQuery(ctx context.Context, query *wire.OpQuery) (*wire.OpReply, error) {
-	if query.FullCollectionName == "admin.$cmd" {
-		switch cmd := query.Query.Command(); cmd {
-		case "ismaster", "isMaster": // both are valid
-			reply := &wire.OpReply{
-				NumberReturned: 1,
-				Documents: []*types.Document{must.NotFail(types.NewDocument(
-					"ismaster", true, // only lowercase
-					// topologyVersion
-					"maxBsonObjectSize", int32(types.MaxDocumentLen),
-					"maxMessageSizeBytes", int32(wire.MaxMsgLen),
-					"maxWriteBatchSize", int32(100000),
-					"localTime", time.Now(),
-					// logicalSessionTimeoutMinutes
-					"connectionId", int32(42),
-					"minWireVersion", common.MinWireVersion,
-					"maxWireVersion", common.MaxWireVersion,
-					"readOnly", false,
-					"ok", float64(1),
-				))},
+	cmd := query.Query.Command()
+	collection := query.FullCollectionName
+
+	// both are valid
+	if (cmd == "ismaster" || cmd == "isMaster") && collection == "admin.$cmd" {
+		return common.IsMaster(ctx, query)
+	}
+
+	// defaults to the database name if supplied on the connection string or $external
+	if cmd == "saslStart" &&
+		(collection == "$external.$cmd" || strings.HasSuffix(collection, "$cmd")) {
+		var emptyPayload types.Binary
+
+		if err := common.SASLStart(ctx, query.Query); err != nil {
+			return nil, lazyerrors.Error(err)
+		}
+
+		if _, err := h.DBPool(ctx); err != nil {
+			var pgErr *pgconn.PgError
+			if errors.As(err, &pgErr) && pgerrcode.IsInvalidAuthorizationSpecification(pgErr.Code) {
+				msg := "FerretDB failed to authenticate you in PostgreSQL:\n" +
+					strings.TrimSpace(pgErr.Error()) + "\n" +
+					"See https://docs.ferretdb.io/security/authentication/ for more details."
+
+				return nil, commonerrors.NewCommandErrorMsgWithArgument(
+					commonerrors.ErrAuthenticationFailed,
+					msg,
+					"payload",
+				)
```

Comment:
Codecov thinks L53-L64 (including successfully authenticating DB) are not covered by tests. Could you add test please?

Reply:
Yeah found that out myself too!